doc/port_auto_neg/port-auto-negotiation-design.md

```diff
@@ -0,0 +1,452 @@
+- adv_interface_types: string value "all" or a list of interface type values defined in IEEE 802.3. For example: "CR4,KR4,SR4", "all".
+
+#### Application DB Enhancements
+
+The change in APP_DB is similar to CONFIG_DB. 3 new fields **adv_speeds**, **interface_type**, **adv_interface_types** will be added to **PORT_TABLE** table:
+
+	; Defines information for port configuration
+	key                     = PORT_TABLE:port_name           ; configuration of the port
+	; field                 = value
+    ...
+	adv_speeds              = STRING                         ; advertised speed list
+	interface_type          = STRING                         ; interface type
+	adv_interface_types     = STRING                         ; advertised interface types
+
+Valid value of the new fields are the same as **PORT** table in CONFIG_DB.
+
+Here is the table to map the fields and SAI attributes:
+| **Parameter**       | **sai_port_attr_t**
+|---------------------|------------------------------------------------|
+| adv_interface_types | SAI_PORT_ATTR_ADVERTISED_INTERFACE_TYPE        |
+| adv_speeds          | SAI_PORT_ATTR_ADVERTISED_SPEED                 |
+| autoneg             | SAI_PORT_ATTR_AUTO_NEG_MODE                    |
+| interface_type      | SAI_PORT_ATTR_INTERFACE_TYPE                   |
+| speed               | SAI_PORT_ATTR_SPEED                            |
+
+#### DB Migrator Enhancements
+
+In current SONiC implementation, if auto negotiation is enabled, it uses the `speed` field as the advertised speeds. Since this feature introduced a new field `adv_speeds`, we need do DB migration to keep backward compatible. For example, the configuration: 
+
+```json
+"Ethernet0": {
+    ...
+    "autoneg": "1",
+    "speed": "100000"
+}
+```
+
+Will be migrated to:
+
+```json
+"Ethernet0": {
+    ...
+    "autoneg": "1",
+    "speed": "100000",
+    "adv_speeds": "100000"
+}
+```
+
+#### SWSS Enhancements
+
+The current SONiC speed setting flow in PortsOrch can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+
+if autoneg == true and speed is set:
+    setPortAdvSpeed(port, speed)
+else if autoneg == false and speed is set:
+    setPortSpeed(port, speed)
+```
+
+The new SONiC speed setting flow can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+
+if autoneg == true:
+    speed_list = vector()
+    if adv_speeds is set:
+        speed_list = adv_speeds
+    else:
+        speed_list = all_supported_speeds
+    setPortAdvSpeed(port, speed_list)
+
+    interface_type_list = vector()
+    if adv_interface_types is set:
+        interface_type_list = adv_interface_types
+    else:
+        interface_type_list = all_supported_interface_types
+    setPortAdvInterfaceType(port, interface_type_list)
+else:
+    if speed is set:
+        setPortSpeed(port, speed)
+    if interface_type is set:
```

Comment:
What if interface_type is not given?

Reply:
This is vendor specified and SAI must keep backward compatible. In that case, we need figure out the current behavior of low level layer and make sure it behaves the same after this feature.